File: source/c/kruskal_mst.c
// gcc -o kruskal_mst kruskal_mst.c
#include <stdio.h>

#define N 5

int uf_parent[N], uf_rank[N];

void uf_init(void) {
    for (int i = 0; i < N; i++) { uf_parent[i] = i; uf_rank[i] = 0; }
}

int uf_find(int x) {
    while (uf_parent[x] != x) {
        uf_parent[x] = uf_parent[uf_parent[x]];
        x = uf_parent[x];
    }
    return x;
}

int uf_union(int a, int b) {
    int ra = uf_find(a), rb = uf_find(b);
    if (ra == rb) return 0;
    if (uf_rank[ra] < uf_rank[rb]) uf_parent[ra] = rb;
    else if (uf_rank[ra] > uf_rank[rb]) uf_parent[rb] = ra;
    else { uf_parent[rb] = ra; uf_rank[ra]++; }
    return 1;
}

typedef struct { int u, v, w; } Edge;

int main(void) {
    Edge edges[] = {{0,1,2},{0,3,6},{1,2,3},{1,3,8},{1,4,5},{2,4,7},{3,4,9}};
    int ne = 7;

    // Insertion sort by weight
    for (int i = 1; i < ne; i++) {
        Edge ke = edges[i]; int j = i - 1;
        while (j >= 0 && edges[j].w > ke.w) { edges[j+1] = edges[j]; j--; }
        edges[j+1] = ke;
    }

    uf_init();

    Edge mst[N-1];
    int mst_len = 0, total = 0;
    for (int i = 0; i < ne && mst_len < N - 1; i++) {
        if (uf_union(edges[i].u, edges[i].v)) {
            mst[mst_len++] = edges[i];
            total += edges[i].w;
        }
    }

    printf("kruskal MST edges:\n");
    for (int i = 0; i < mst_len; i++) {
        printf("  %d - %d (weight %d)\n", mst[i].u, mst[i].v, mst[i].w);
    }
    printf("total weight: %d\n", total);
    return 0;
}

File: source/javascript/kruskal_mst.js
// Kruskal's Minimum Spanning Tree with Union-Find
const ufParent = [];
const ufRank = [];

function ufInit(n) {
    for (let i = 0; i < n; i++) { ufParent.push(i); ufRank.push(0); }
}

function ufFind(x) {
    while (ufParent[x] !== x) {
        ufParent[x] = ufParent[ufParent[x]];
        x = ufParent[x];
    }
    return x;
}

function ufUnion(a, b) {
    const ra = ufFind(a), rb = ufFind(b);
    if (ra === rb) return false;
    if (ufRank[ra] < ufRank[rb]) ufParent[ra] = rb;
    else if (ufRank[ra] > ufRank[rb]) ufParent[rb] = ra;
    else { ufParent[rb] = ra; ufRank[ra]++; }
    return true;
}

const n = 5;
const edges = [
    { u: 0, v: 1, w: 2 }, { u: 0, v: 3, w: 6 }, { u: 1, v: 2, w: 3 },
    { u: 1, v: 3, w: 8 }, { u: 1, v: 4, w: 5 }, { u: 2, v: 4, w: 7 },
    { u: 3, v: 4, w: 9 }
];

// Sort by weight
edges.sort((a, b) => a.w - b.w);
ufInit(n);

const mstEdges = [];
let total = 0;
for (const edge of edges) {
    if (ufUnion(edge.u, edge.v)) {
        mstEdges.push(edge);
        total += edge.w;
    }
}

console.log("kruskal MST edges:");
for (const e of mstEdges) console.log(`  ${e.u} - ${e.v} (weight ${e.w})`);
console.log(`total weight: ${total}`);
